### ISSUE TITLE
Epic: Multiple file output

### ISSUE THREAD

**Issue by RSuter** [RSuter/NJsonSchema#514]
Generate multiple output artifacts instead of one output file (code generators). 

Artifact properties: 

- Name (e.g. Person)
- Code
- Filename (e.g. Person.cs)
- Dependencies (list of artifact dependencies, e.g. [Address, Skill])

**Comment by RSuter** [comment]
All generators inherit from GeneratorBase which has these two methods: 

```
        /// <summary>Generates the type from the schema and all types from the resolver.</summary>
        /// <param name="schema">The schema</param>
        /// <param name="typeNameHint">The type name hint.</param>
        /// <returns>The code.</returns>
        public CodeArtifactCollection GenerateTypes(JsonSchema4 schema, string typeNameHint)
        {
            _resolver.Resolve(schema, false, typeNameHint); // register root type
            return GenerateTypes();
        }

        /// <summary>Generates the the whole file containing all needed types.</summary>
        /// <returns>The code</returns>
        public string GenerateFile(JsonSchema4 schema, string typeNameHint)
        {
            var collection = GenerateTypes(schema, typeNameHint);
            return GenerateFile(collection);
        }
```

so instead of using GenerateFile you can call GenerateTypes and you will get a collection of artifacts (i.e. code files)

**Comment by blueboxes** [comment]
Great, thank you.

**Comment by blueboxes** [comment]
The SwaggerToCSharpClientGenerator does not expose this, and I cannot see the relationship between this and CSharpGenerator that does?

**Comment by RSuter** [comment]
At the moment it's just implemented in NJS. It's not yet fully implemented in NSwag...

**Comment by blueboxes** [comment]
ok, thanks for your help

**Comment by MarcusSpecht** [comment]
yes yes yes... i would love to have my "models" in seperate files. For C# and Typescript 💃 that would be awesome

**Comment by lecoque** [comment]
@RSuter is this implemented in NSwag now?

Is there any solution for this?

**Comment by RSuter** [comment]
Nope, not yet implemented in NSwag - first the generators have to updated and then the whole tooling, CLI and UI - lots of work and I personally dont need it. This is why I prioritize other stuff (e.g. ASP.NET Core 2.1 support, etc.)

**Comment by larslarsson** [comment]
Why is this closed, it doesn't exist in the code base, but can be achived inheriting the generator & feeding it individual/or groups of artifacts. Or am I missing something?

**Comment by RicoSuter** [comment]
Its implemented only in njs (here) but must be used in nswag (not yet)

**Comment by larslarsson** [comment]
How do you use it in NJsonSchema? Did I miss that in the replies?
The looping does not generate complete class files unless one override CSharpGenerator in my case (using the protected GenerateFile(ArtifactCollection collection).

**Comment by skotl** [comment]
I haven't done anything with this, but it may be of use to some people.
There are protected methods in `CSharpClientGenerator` called:

`GenerateDtoTypes()` 
`GenerateClientTypes()`

Both return an `IEnumerable<CodeArtifact>` which exposes (at least) three useful properties:
`Type`, `Code` and `TypeName`.

If you create a new generator class that inherits from `CSharpClientGenerator` then you can provide a public method to call the underlying generator, return the code artefacts and then write them to individual files named after `TypeName`. Something like this:

```public class g2 : CSharpClientGenerator
    {
        ...
        public IEnumerable<CodeArtifact> GenerateDtoTypes()
        {
            return base.GenerateDtoTypes();
        }
    }

    ...

            var generator = new g2(doc, settings);
            foreach (var t in generator.GenerateDtoTypes())
            {
                WriteMyTypeToFile(t.TypeName, t.Code);
            }
```

**Tons of caveats around the underlying classes and methods may change...**

**Comment by RicoSuter** [comment]
Yes, this is how this feature will be implemented eventually
